### Segmenting and Clustering Neighborhoods in Toronto

In [27]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
!pip install folium
import folium # map rendering library

print('Libraries imported.')

Solving environment: | 
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/linux-64::anaconda==5.3.1=py37_0
  - defaults/linux-64::astropy==3.0.4=py37h14c3975_0
  - defaults/linux-64::bkcharts==0.2=py37_0
  - defaults/linux-64::blaze==0.11.3=py37_0
  - defaults/linux-64::bokeh==0.13.0=py37_0
  - defaults/linux-64::bottleneck==1.2.1=py37h035aef0_1
  - defaults/linux-64::dask==0.19.1=py37_0
  - defaults/linux-64::datashape==0.5.4=py37_1
  - defaults/linux-64::mkl-service==1.1.2=py37h90e4bf4_5
  - defaults/linux-64::numba==0.39.0=py37h04863e7_0
  - defaults/linux-64::numexpr==2.6.8=py37hd89afb7_0
  - defaults/linux-64::odo==0.5.1=py37_0
  - defaults/linux-64::pytables==3.4.4=py37ha205bf6_0
  - defaults/linux-64::pytest-arraydiff==0.2=py37h39e3cac_0
  - defaults/linux-64::pytest-astropy==0.4.0=py37_0
  - defaults/linux-64::pytest-doctestplus==0.1.3=py37_0
  - defaults/linux-64::pywavelets==1.0.0=py37h

In [28]:
# To build the code to scrape the following Wikipedia page
#$ pip install beautifulsoup4
#$ easy_install beautifulsoup4
from bs4 import BeautifulSoup
url=  requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
#with open (url) as html_file:
soup = BeautifulSoup(url.text,'lxml')
#print(soup.prettify())

In [29]:
import csv
csv_file=open('canada1.csv','w')
csv_writer=csv.writer(csv_file)
csv_writer.writerow(['Postcode', 'Borough', 'Neighbourhood'])
for tr in soup.find_all('tr')[1:]:
    tds = tr.find_all('td')
    if len(tds)==3:
        Postcode=tds[0].text
        Borough=tds[1].text
        Neighbourhood=tds[2].text
    #print(Postcode, Borough, Neighbourhood)
    csv_writer.writerow([Postcode, Borough, Neighbourhood])
csv_file.close()

### Load table from a clean canada csv data

In [30]:
#load canada data from CSV
import pandas as pd
df_cd=pd.read_csv('canada1.csv')
print('Data loaded')

Data loaded


In [31]:
df_cd.head()

Postcode           Borough       Neighbourhood
0      M1A      Not assigned      Not assigned\n
1      M2A      Not assigned      Not assigned\n
2      M3A        North York         Parkwoods\n
3      M4A        North York  Victoria Village\n
4      M5A  Downtown Toronto      Harbourfront\n

In [32]:
df_cd[df_cd.columns] = df_cd.apply(lambda x: x.str.strip('\n'))
df_cd.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

### I Ignore cells with a borough that is Not assigned

In [33]:
# excluding borough that is Not assigned
df_ab=df_cd[df_cd.Borough!='Not assigned']
df_ab.head()
df_ab.shape

(211, 3)

### Where more than one neighborhood exist in one postal code area are combine in a single row separated by comma
For example, in the table on the Wikipedia page,  M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. 

In [34]:
df_ab.set_index(['Postcode','Borough'],inplace=True)
res = df_ab.groupby(level=['Postcode','Borough'], sort=False).agg( ','.join)
#res.dtypes

In [35]:
res.reset_index() # to remove set index

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
5        M9A         Etobicoke   
6        M1B       Scarborough   
7        M3B        North York   
8        M4B         East York   
9        M5B  Downtown Toronto   
10       M6B        North York   
11       M9B         Etobicoke   
12       M1C       Scarborough   
13       M3C        North York   
14       M4C         East York   
15       M5C  Downtown Toronto   
16       M6C              York   
17       M9C         Etobicoke   
18       M1E       Scarborough   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
21       M6E              York   
22       M1G       Scarborough   
23       M4G         East York   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
26       M1H       Scarborough   
27       M2H        North York   
28       M3H        North York   
29       M4H         East York   
30       M5H  Downtown Toronto   
31       M6H      West Toronto   
32       M1J       Scarborough   
33       M2J        North York   
34       M3J        North York   
35       M4J         East York   
36       M5J  Downtown Toronto   
37       M6J      West Toronto   
38       M1K       Scarborough   
39       M2K        North York   
40       M3K        North York   
41       M4K      East Toronto   
42       M5K  Downtown Toronto   
43       M6K      West Toronto   
44       M1L       Scarborough   
45       M2L        North York   
46       M3L        North York   
47       M4L      East Toronto   
48       M5L  Downtown Toronto   
49       M6L        North York   
50       M9L        North York   
51       M1M       Scarborough   
52       M2M        North York   
53       M3M        North York   
54       M4M      East Toronto   
55       M5M        North York   
56       M6M              York   
57       M9M        North York   
58       M1N       Scarborough   
59       M2N        North York   
60       M3N        North York   
61       M4N   Central Toronto   
62       M5N   Central Toronto   
63       M6N              York   
64       M9N              York   
65       M1P       Scarborough   
66       M2P        North York   
67       M4P   Central Toronto   
68       M5P   Central Toronto   
69       M6P      West Toronto   
70       M9P         Etobicoke   
71       M1R       Scarborough   
72       M2R        North York   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
76       M7R       Mississauga   
77       M9R         Etobicoke   
78       M1S       Scarborough   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
82       M1T       Scarborough   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
85       M1V       Scarborough   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
88       M8V         Etobicoke   
89       M9V         Etobicoke   
90       M1W       Scarborough   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
93       M8W         Etobicoke   
94       M9W         Etobicoke   
95       M1X       Scarborough   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Harbourfront,Regent Park  
3                      Lawrence Heights,Lawrence Manor  
4                                         Not assigned  
5                                     Islington Avenue  
6                                        Rouge,Malvern  
7          

### A cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. 
So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.

In [36]:
res.loc[res['Neighbourhood'] =='Not assigned']=res[res['Neighbourhood'] == 'Not assigned'].index.values[0][1]
res.head()

Neighbourhood
Postcode Borough                                          
M3A      North York                              Parkwoods
M4A      North York                       Victoria Village
M5A      Downtown Toronto         Harbourfront,Regent Park
M6A      North York        Lawrence Heights,Lawrence Manor
M7A      Queen's Park                         Queen's Park

In [37]:
df=res.reset_index()
df.head()

Postcode           Borough                    Neighbourhood
0      M3A        North York                        Parkwoods
1      M4A        North York                 Victoria Village
2      M5A  Downtown Toronto         Harbourfront,Regent Park
3      M6A        North York  Lawrence Heights,Lawrence Manor
4      M7A      Queen's Park                     Queen's Park

### Using the .shape method to print the number of rows of your dataframe.

In [38]:
df.shape

(103, 3)

Now, you can see that we have 103 rows and 3 columns

In [39]:
df["Address"]=df['Borough'] +", " +df['Postcode'] 
df.head()

Postcode           Borough                    Neighbourhood  \
0      M3A        North York                        Parkwoods   
1      M4A        North York                 Victoria Village   
2      M5A  Downtown Toronto         Harbourfront,Regent Park   
3      M6A        North York  Lawrence Heights,Lawrence Manor   
4      M7A      Queen's Park                     Queen's Park   

                 Address  
0        North York, M3A  
1        North York, M4A  
2  Downtown Toronto, M5A  
3        North York, M6A  
4      Queen's Park, M7A

### Generate list of urls using my API key first

In [49]:
url2 = [] #generate list of urls
for i in range(0, len(df)):
    url = 'https://maps.googleapis.com/maps/api/geocode/json?address=+Toronto+Ontario&key=MY-API-KEY'.format(df['Address'][i])
    url2.append(url)
 

#### Now loop through the url to get latitude and longitude from google API

In [50]:
df['Latitude']=0
df['Longitude']=0

for i, url in enumerate(url2):
    r = requests.get(url2[i], params=params)
    results = r.json()['results']
    if results:
        location = results[0]['geometry']['location']
    else:
        print ("no results")
    df['Latitude'].iloc[i] = location['lat']
    df['Longitude'].iloc[i]= location['lng']

/home/jupyterlab/conda/lib/python3.6/site-packages/pandas/core/indexing.py:190: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


In [51]:
df.head(50)

Postcode           Borough  \
0       M3A        North York   
1       M4A        North York   
2       M5A  Downtown Toronto   
3       M6A        North York   
4       M7A      Queen's Park   
5       M9A         Etobicoke   
6       M1B       Scarborough   
7       M3B        North York   
8       M4B         East York   
9       M5B  Downtown Toronto   
10      M6B        North York   
11      M9B         Etobicoke   
12      M1C       Scarborough   
13      M3C        North York   
14      M4C         East York   
15      M5C  Downtown Toronto   
16      M6C              York   
17      M9C         Etobicoke   
18      M1E       Scarborough   
19      M4E      East Toronto   
20      M5E  Downtown Toronto   
21      M6E              York   
22      M1G       Scarborough   
23      M4G         East York   
24      M5G  Downtown Toronto   
25      M6G  Downtown Toronto   
26      M1H       Scarborough   
27      M2H        North York   
28      M3H        North York   
29      M4H         East York   
30      M5H  Downtown Toronto   
31      M6H      West Toronto   
32      M1J       Scarborough   
33      M2J        North York   
34      M3J        North York   
35      M4J         East York   
36      M5J  Downtown Toronto   
37      M6J      West Toronto   
38      M1K       Scarborough   
39      M2K        North York   
40      M3K        North York   
41      M4K      East Toronto   
42      M5K  Downtown Toronto   
43      M6K      West Toronto   
44      M1L       Scarborough   
45      M2L        North York   
46      M3L        North York   
47      M4L      East Toronto   
48      M5L  Downtown Toronto   
49      M6L        North York   

                                        Neighbourhood                Address  \
0                                           Parkwoods        North York, M3A   
1                                    Victoria Village        North York, M4A   
2                            Harbourfront,Regent Park  Downtown Toronto, M5A   
3                     Lawrence Heights,Lawrence Manor        North York, M6A   
4                                        Queen's Park      Queen's Park, M7A   
5                                    Islington Avenue         Etobicoke, M9A   
6                                       Rouge,Malvern       Scarborough, M1B   
7                                     Don Mills North        North York, M3B   
8                      Woodbine Gardens,Parkview Hill         East York, M4B   
9                             Ryerson,Garden District  Downtown Toronto, M5B   
10                                          Glencairn        North York, M6B   
11  Cloverdale,Islington,Martin Grove,Princess Gar...         Etobicoke, M9B   
12               Highland Creek,Rouge Hill,Port Union       Scarborough, M1C   
13                    Flemingdon Park,Don Mills South        North York, M3C   
14                                   Woodbine Heights         East York, M4C   
15                                     St. James Town  Downtown Toronto, M5C   
16                                 Humewood-Cedarvale              York, M6C   
17  Bloordale Gardens,Eringate,Markland Wood,Old B...         Etobicoke, M9C   
18                    Guildwood,Morningside,West Hill       Scarborough, M1E   
19                                        The Beaches      East Toronto, M4E   
20                                        Berczy Park  Downtown Toronto, M5E   
21                                Caledonia-Fairbanks              York, M6E   
22                                             Woburn       Scarborough, M1G   
23                                            Leaside         East York, M4G   
24                                 Central Bay Street  Downtown Toronto, M5G   
25                                           Christie  Downtown Toronto, M6G   
26                                          Cedarbrae       Scarborough, M1H   
27                                  Hillcrest Village        North York, M2H   

I write to csv so as not to run the for loop again

In [52]:
df.to_csv('canada_cd.csv')